man/man8/ixtimezone.8
.de Sp
.if n .sp
.if t .sp 0.4
..
.TH IXTIMEZONE 8
.SH NAME
ixtimezone \- allows Unix timezone-handling to be used with ixemul.library
.SH SYNOPSIS
.B ixtimezone
option
.br
.SH DESCRIPTION
.I Ixtimezone
uses the
.B TZ
environment variable to determine the offset in seconds between Universal
Coordinated Time (aka Greenwich Mean Time) and the local timezone. This
offset is stored in a field in the ixemul library base. The offset is also
written to the environment variable
.BR IXGMTOFFSET .
The 
.I ixemul.library
is now able to correctly determine the local time and
GMT.
.LP
The
.B TZ
environment variable should contain the name of one of the files in the
.B /etc/zoneinfo
directory. For example, I live in the Netherlands, so my
.B TZ
is set to
.BR Europe/Amsterdam .
The
.B TZ
environment variable
.I must
be set before using
.IR ixtimezone ,
otherwise
.I ixtimezone
wouldn't know what the local timezone is, and therefore it wouldn't be able
to compute the offset from GMT.
.LP
You should put
.I ixtimezone
in your
.B startup-sequence
to ensure that it is called every time you start your computer. After
.I ixtimezone
has been executed, you can set
.B TZ
to whatever value you want. For example, to see what time it is in Japan,
set
.B TZ
to
.B Japan
and type
.I ixtimezone
.IR \-test .
But don't use any other option, as the other options set the GMT offset based on the
current
.B TZ
value, which is
.BR Japan !
.LP
Note that you need to call this tool at least twice a year, when
Daylight Saving Time comes in effect and when it ends.
.SH OPTIONS
.TP
.B \-test
Print the local time and GMT (according to 
.IR ixemul.library )
to standard output.
Use this to test the current settings. It doesn't affect anything else.
.TP
.B \-get\-offset
Calculate the difference between local time and GMT in seconds. The
resulting offset is stored in the library base of
.I ixemul.library
and is also written to the global environment variable
.BR IXGMTOFFSET .
.Sp
If the
.B IXGMTOFFSET
environment variable didn't exist or the offset stored in the original
environment variable differed from the new offset (which happens when
Daylight Saving Time starts or ends), then a copy of
.B IXGMTOFFSET
is written to the
.B ENVARC:
directory.
.Sp
Since
.I ixemul.library
reads the
.B IXGMTOFFSET
environment variable just after it is loaded into memory, it is
sufficient to use
.I ixtimezone
only when Daylight Saving Time starts or ends. Provided, of course, that
the contents of the
.B ENVARC:
directory is copied to 
.B ENV:
before
.I ixemul.library
is loaded.
.TP
.B \-check\-dst
This option does the same as
.BR \-get\-offset ,
but it also detects if Daylight Saving Time has started or ended. If so, it
automatically sets the Amiga clock to the correct time. So you no longer
need to change the time manually twice a year.
.TP
.B \-patch\-resource
This option does the same as
.BR \-get\-offset ,
but it also patches the
.IR battclock.resource .
This is most useful if you also run the Unix operating system on your Amiga
(NetBSD or Linux, for example). Unix requires that the clock is set to
Greenwich Mean Time, while the Amiga requires that the clock is set to
local time. By patching the 
.I battclock.resource
one can set the Amiga clock, which is managed by the
.IR battclock.resource ,
to GMT time. The patch ensures that whenever the Amiga wants to
obtain the time, the GMT time is converted into local time. There is also
no longer any need to change the Amiga clock when Daylight Saving Time starts or
ends, since the battery clock is set to GMT.
.Sp
.I ixtimezone
detects if you try to patch the
.I battclock.resource
a second time. It won't patch the resource again, but it will check if the
current GMT offset is still the same as when the patch was originally
installed. If it has changed, the
.I battclock.resource
offset will also be changed.
.TP
.B \-remove\-patch
Removes the 
.I battclock.resource
patch. It checks if the patch was really installed before removing it.
This option doesn't do anything else.
.SH FILE
/etc/zoneinfo	standard directory used for created files
.SH HISTORY
The original idea for the 
.B \-patch\-resource
option came from the
.I unixclock
utility written by Geert Uytterhoeven.
.I unixclock
is available on Aminet.
.SH "SEE ALSO"
The time functions, the timezone databases and the
.IR zic (8)
utility for compiling the timezone files in the
.B /etc/zoneinfo
directory are available from
.I elsie.nci.nih.gov
in the
.B /pub
directory.
.SH AUTHOR
Hans Verkuil
.RI ( hans@wyst.hobby.nl )


NEWS
============
VERSION 48.2
============

Removed all pOS code and references.

Fixed error reporting in __extend_file().

Fixed potential problem of reading from freed memory, and enforcer hits
caused from running non-ixemul programs under ixemul control (ie. make).
Patches supplied by Emmanuel Lesueur.

============
VERSION 48.1
============

Interim release.

============
VERSION 48.0
============

Fixed a bug in ftruncate reported by Niels Froehling (incorrect
interpretation of the SetFileSize return value).

Fixed a malloc bug (it no longer accepts sizes < 0).

Added a few things for Holger Jakob's ppc ixemul support.

p.OS is no longer supported. While 99% of the code is still there
it is no longer compiled, nor supported. Binaries compiled for
ixemul 48.0 will no longer work under a p.OS system, nor do you
need the p.OS headers anymore to compile ixemul.library.

============
VERSION 47.3
============

Fixed a bug in getUserPasswd where certain strings weren't properly
0-terminated.

Fixed a bug in setpriority and getpriority: these functions always failed
due to an wrong test.

Fixed a bug in ctype.h: all macros (e.g. isspace and friends) failed to
work properly with character code >= 128. Now cast everything to unsigned
char first.

Applied patch from David Zaroski to uname.c. Improved OS version number
reporting.

============
VERSION 47.2
============

Fixed a bug in handling of "/". Thanks to Norbert Pueschel for providing me
with a fix.

Updated ixprefs from Kriton Kyrimis.

Fixed problems with vfork() and sockets when the program was started from
inetd.

============
VERSION 47.1
============

Fixed ixtrace bug that made it impossible to set the stacksize.

Fixed bug in crt0.c which caused a crash under p.OS.

============
VERSION 47.0
============

It was possible to kill a process with a SIGKILL signal when that process
was already in the exit() function: this can cause major crashes.

Fixed more bugs in wait4(): the list of child processes could contain
processes whose 'getuser()' pointer was NULL, causing enforcer hits and
general mayhem.

Signalling processes that were asleep waiting for some event would not wake
these processes.

Fixed a bug in wait4: if there are no longer any childs available that
meet the criteria of wait4, then wait4 should exit. In previous versions
this test was only done for the zombie processes, and not for all other
childs.

Updated timezone sources and databases.

Kriton Kyrimis supplied patches to port ixprefs to p.OS.

David Zaroski supplied patches to upgrade ixtrace to this ixemul version.

vfork() will now copy any open sockets to the child, previously the
sockets inherited by the child from the parent were always invalid.

Added Unix Sys V semaphore and message support (sem* and msg* functions).

Added low-level mutex (semaphore) functions: ix_mutex_lock(),
ix_mutex_attempt_lock() and ix_mutex_unlock(). These functions are
wrappers for the AmigaOS or p.OS kernel semaphore functions. See ix.h for
more info.

Fixed several nasty bugs in the shm support code.

Added a new checkbox to ixprefs: Enable MuFS support. By default the
support for the Multi-user File System is turned off. One reason for adding
this checkbox is that apparently the MuFS support isn't bug-free.

ade -> gg

Added workaround for AFS bug with symlink("", "").

The resolv.conf keyword scan is now case-insensitive: the AmiTCP
resolv.conf file is uppercase, while NetBSD assumes lowercase.

Applied patches from Norbert Pueschel fixing argv[0] when starting a
script: it was the name of the scripts, now it is the name of the
interpreter. This conforms to NetBSD and fixes problems with dld.

ttyname() in libc.a now always returns "/dev/tty" if the device is a tty
and NULL otherwise.

"#!shell arg" handling has been changed to be consistent with NetBSD: all
arguments after the shell name are now passed as the first argument to the
shell.

munmap() now calls msync().

Ixemul now fully supports pOS.

chown() will succeed if the device doesn't support this action instead of
reporting an error.

The -enforcer flag of gdb now also supports the 68060.

Improved getppid and added hstrerror thanks to Norbert Pueschel.

Added ix_obtain_socket and ix_release_socket thanks to Jeff Shepherd.
Added __flush_cache for gcc trampoline handling.
Added ix_flush_insn_cache, ix_flush_caches and ix_flush_data_cache for OS
independent cache flushing.
Added ix_get_first_segment and ix_get_next_segment.
Added ix_select.
Added isascii, isiso, toascii and toiso.
Added ix_get_long, ix_set_long.

Removed aselect.
Removed CreatePort, CreateExtIO, CreateStdIO, CreateTask.
Removed DeletePort, DeleteExtIO, DeleteStdIO, DeleteTask.
Removed ix_get_vars.
Removed _wb_parse, _cli_parse and _main.

Renamed achmod to ix_chmod.
Renamed ix_get_vars2 to ix_get_vars.
Renamed vfork2 to ix_vfork.
Renamed vfork_resume to ix_vfork_resume.
Renamed __amiga_filehandle to ix_filehandle.
Renamed tracecntl to ix_tracecntl.
Renamed ix_amiga.h to ix.h.

============
VERSION 46.1
============

Renamed __os to ix_os as a preparation for ixemul-47.0. Also renamed
expand_cmd_line to ix_expand_cmd_line and default_wb_window to
ix_default_wb_window.

============
VERSION 46.0
============

Copied basic pOS support from ixemul-47.0 especially for GeekGadgets 2. The
ixemul.library does *not* support pOS, you have to use ixemul-47.0 for
that, but all programs compiled with the new libc.a and *crt0.o will also
run on pOS using ixemul-47.0. Unless Amiga kernel functions are used, of
course. See README.pOS for more information.

Added ipcs and ipcrm to manage the shm functionality.

Upgraded db to version 1.85.

Added an ixprefs option to show the stack usage of all the programs that
use ixemul.library.

chmod() now also checks if someone else has the same file open, and makes
sure that when that file is closed the same mode flags are written instead
of the flags specified when the file was opened.


utils/ixtimezone.c
/*
 *  Written by Hans Verkuil (hans@wyst.hobby.nl)
 *
 *  battclock.resource patch idea was shamelessly stolen from the unixclock
 *  utility written by Geert Uytterhoeven. unixclock is available on Aminet.
 */

#include <stdlib.h>
#include <stdio.h>
#include <time.h>

#include <ixemul.h>
#include <exec/memory.h>
#include <clib/exec_protos.h>
#include <dos/var.h>

#include <sys/time.h>
#include <proto/exec.h>
#include <proto/dos.h>
#include <proto/battclock.h>

#define LVOReadBattClock	(-12)
#define LVOWriteBattClock	(-18)

/* Flags. Currently only bit 0 is used to tell whether Daylight Saving Time
   is in effect or not. However, ixtimezone only sets this flag, but it doesn't
   test it. And neither does ixemul.library. In fact, the library completely
   ignores the flag field. */

#define DST_ON	0x01

typedef struct
{
  long          offset;
  unsigned char flags;
} ixtime;

struct Library *BattClockBase;

char VERSION[] = "\000$VER: ixtimezone 1.0 (10.11.95)";

/* battclock.resource patch code */
asm("
	.text
	.globl _NewReadBattClock
	.globl _NewWriteBattClock
	.globl _OldReadBattClock
	.globl _OldWriteBattClock
	.globl _GMTOffset
	.globl _EndOfPatch

_NewReadBattClock:
	.int	0x207a0014, 0x4e9090ba, 0x00164e75

/*	Since the GNU assembler is currently unable to properly compile
	PC-relative code, I'm using the hex-code directly. As soon as the
	assembler can handle PC-relative code, the line above should be
	replaced by:

	move.l		_OldReadBattClock(pc),a0
	jsr		(a0)
	sub.l		_GMTOffset(pc),d0
	rts
*/

_NewWriteBattClock:
	.int	0xd0ba0010, 0x207a0008
	.short	0x4ed0

/*	Since the GNU assembler is currently unable to properly compile
	PC-relative code, I'm using the hex-code directly. As soon as the
	assembler can handle PC-relative code, the line above should be
	replaced by:

	add.l		_GMTOffset(pc),d0
	move.l		_OldWriteBattClock(pc),a0
	jmp		(a0)
*/

_OldReadBattClock:
	.int		0
_OldWriteBattClock:
	.int		0
_GMTOffset:
	.int		0
_EndOfPatch:
");

extern char NewReadBattClock;
extern char NewWriteBattClock;
extern char OldReadBattClock;
extern char OldWriteBattClock;
extern long GMTOffset; 
extern long EndOfPatch;

static ixtime *read_ixtime(void)
{
  static ixtime t;
  char buf[10];
  
  /*
   * Read the GMT offset. This environment variable is 5 bytes long. The
   * first 4 form a long that contains the offset in seconds and the fifth
   * byte contains flags.
   */
  if (GetVar("IXGMTOFFSET", buf, 6, GVF_BINARY_VAR) == 5 && IoErr() == 5)
  {
    memcpy(&t.offset, buf, 4);
    t.flags = buf[4];
    return &t;
  }
  return NULL;
}

static void create_ixtime(ixtime *t, char *pathname)
{
  FILE *f = fopen(pathname, "w");
  
  if (f)
  {
    fwrite(t, 5, 1, f);
    fclose(f);
  }
}

static void write_ixtime(ixtime *t, int write_also_to_envarc)
{
  ix_set_gmt_offset(t->offset);
  create_ixtime(t, "/ENV/IXGMTOFFSET");
  if (write_also_to_envarc)
    create_ixtime(t, "/ENVARC/IXGMTOFFSET");
}

static void set_clock(long offset)
{
  struct timeval tv;
  
  gettimeofday(&tv, NULL);
  tv.tv_sec += offset;
  settimeofday(&tv, NULL);
}

static void reset_clock(void)
{
  struct timeval tv;
  
  tv.tv_usec = 0;
  tv.tv_sec = ReadBattClock();
  tv.tv_sec += (8*365+2) * 24 * 3600 + ix_get_gmt_offset();
  settimeofday(&tv, NULL);
}

static long *get_function_addr(void)
{
  return *((long **)((char *)BattClockBase + LVOReadBattClock + 2));
}

static void patch_batt_resource(long offset)
{
  char *mem;
  long size = (long)&EndOfPatch - (long)&NewReadBattClock;
  long mem_offset;
  long *oldread, *oldwrite, *gmt;

  BattClockBase = (struct Node *)OpenResource("battclock.resource");
  if (*(gmt = get_function_addr()) == 0x207a0014)
  {
    printf("battclock.resource was already patched.\n");
    gmt = (long *)(((char *)&GMTOffset) + (long)((char *)gmt - &NewReadBattClock));
    if (*gmt != offset)
    {
      *gmt = offset;
      reset_clock();
    }
    return;  /* already patched */
  }
  GMTOffset = offset;
  mem = AllocMem(size, MEMF_PUBLIC);
  mem_offset = mem - &NewReadBattClock;
  memcpy(mem, &NewReadBattClock, size);
  CacheClearE(mem, size, CACRF_ClearI);  /* clear instruction cache */
  oldread = (long *)(&OldReadBattClock + mem_offset);
  oldwrite = (long *)(&OldWriteBattClock + mem_offset);
  Disable();
  *oldread = (long)SetFunction((struct Library *)BattClockBase, LVOReadBattClock, (void *)(&NewReadBattClock + mem_offset));
  *oldwrite = (long)SetFunction((struct Library *)BattClockBase, LVOWriteBattClock, (void *)(&NewWriteBattClock + mem_offset));
  Enable();
  reset_clock();
  printf("patched battclock.resource.\n");
}

static void remove_patch(void)
{
  long *p, mem_offset, *oldread, *oldwrite;
  long size = (long)&EndOfPatch - (long)&NewReadBattClock;

  BattClockBase = (struct Node *)OpenResource("battclock.resource");
  if (*(p = get_function_addr()) != 0x207a0014)
  {
    printf("battclock.resource wasn't patched.\n");
    exit(0);  /* not patched */
  }
  mem_offset = (char *)p - &NewReadBattClock;
  oldread = (long *)(&OldReadBattClock + mem_offset);
  oldwrite = (long *)(&OldWriteBattClock + mem_offset);
  Disable();
  SetFunction((struct Library *)BattClockBase, LVOReadBattClock, (void *)*oldread);
  SetFunction((struct Library *)BattClockBase, LVOWriteBattClock, (void *)*oldwrite);
  Enable();
  FreeMem(p, size);
  reset_clock();
  printf("removed battclock.resource patch.\n");
  exit(0);
}

static void test(void)
{
  time_t t;
  
  time(&t);
  printf("GMT:   %s", asctime(gmtime(&t)));
  printf("Local: %s", asctime(localtime(&t)));
  exit(0);
}

static void usage(void)
{
  fprintf(stderr, "Usage: ixtimezone <option>
Where <option> is one of:

-test		Show GMT and localtime
-get-offset	Get GMT offset and patch ixemul.library
-check-dst	As -get-offset, but also automatically adjust the Amiga
		time if Daylight Saving Time has gone in effect (or vice
		versa)
-patch-resource	As -get-offset, but also patch the battclock.resource
-remove-patch	Remove the battclock.resource patch\n");
  exit(1);
}

int main(int argc, char **argv)
{
  struct tm *local_tm, *gmt_tm;
  int local_hms, gmt_hms;
  time_t t, local_t, gmt_t;
  ixtime *old, new;
  int set_the_clock = 0, patch_resource = 0;
  int write_to_envarc = 0;

  if (argc != 2)
    usage();

  if (!strcmp(argv[1], "-test"))
    test();
  else if (!strcmp(argv[1], "-check-dst"))
    set_the_clock = 1;
  else if (!strcmp(argv[1], "-patch-resource"))
    patch_resource = 1;
  else if (!strcmp(argv[1], "-remove-patch"))
    remove_patch();
  else if (strcmp(argv[1], "-get-offset"))
    usage();

  /*
   * Get current time, both GMT and local.  
   * We don't care if these values are correct or not, we are only interested
   * in the difference between the two.
   */

  time(&t);
  gmt_tm = gmtime(&t);
  gmt_hms = gmt_tm->tm_hour * 3600 + gmt_tm->tm_min * 60 + gmt_tm->tm_sec;
  local_tm = localtime(&t);
  local_hms = local_tm->tm_hour * 3600 + local_tm->tm_min * 60 + local_tm->tm_sec;
  new.flags = (local_tm->tm_isdst ? DST_ON : 0);
  new.offset = 0;
  if (gmt_hms != local_hms)
  {
    /* They are not the same. So compute the difference between them */

    local_tm->tm_isdst = 0;     /* don't let these values influence the result! */
    local_tm->tm_zone = NULL;
    local_tm->tm_gmtoff = 0;
    local_t = mktime(local_tm);
    gmt_tm = gmtime(&t);
    gmt_tm->tm_isdst = 0;
    gmt_tm->tm_zone = NULL;
    gmt_tm->tm_gmtoff = 0;
    gmt_t = mktime(gmt_tm);
    new.offset = gmt_t - local_t;     /* obtain the difference */
  }

  old = read_ixtime();
  if (old == NULL || old->offset != new.offset)
  {
    write_to_envarc = 1;
    if (set_the_clock && old)
      set_clock(old->offset - new.offset);
  }
  write_ixtime(&new, write_to_envarc);
  if (patch_resource)
    patch_batt_resource(new.offset);
}
